dev-tools/mage/benchmark.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Run execute the go benchmark tests for this repository, define OUTPUT to write results into a file
+func (Benchmark) Run(ctx context.Context) error {
+	mg.Deps(Benchmark.Deps)
+	log.Println(">> go Test: Benchmark")
+	outputFile := os.Getenv("OUTPUT")
+	benchmarkCountOverride := os.Getenv("BENCH_COUNT")
+	if benchmarkCountOverride != "" {
+		var overrideErr error
+		benchmarkCount, overrideErr = strconv.Atoi(benchmarkCountOverride)
+		if overrideErr != nil {
+			return fmt.Errorf("failed to parse BENCH_COUNT, verify that you set the right value: , %w", overrideErr)
+		}
+	}
+	projectPackages, er := gotool.ListProjectPackages()
+	if er != nil {
+		return fmt.Errorf("failed to list package dependencies: %w", er)
+	}
+	cmdArg := fmt.Sprintf("test -count=%d -bench=Bench -run=Bench", benchmarkCount)
+	cmdArgs := strings.Split(cmdArg, " ")
+	for _, pkg := range projectPackages {
+		cmdArgs = append(cmdArgs, filepath.Join(pkg, "/..."))
+	}
+	_, err := runCommand(ctx, nil, "go", outputFile, cmdArgs...)
+
+	var goTestErr *exec.ExitError
+	switch {
+	case goTestErr == nil:
+		return nil
+	case errors.As(err, &goTestErr):
+		return fmt.Errorf("failed to execute go test -bench command: %w", err)
+	default:
+		return fmt.Errorf("failed to execute go test -bench command %w", err)
+	}
+}
+
+// Diff compare 2 benchmark outputs, Required BASE variable for parsing results, define NEXT to compare base with next and optional OUTPUT to write to file
+func (Benchmark) Diff(ctx context.Context) error {
+	mg.Deps(Benchmark.Deps)
+	log.Println(">> running: benchstat")
+	outputFile := os.Getenv("OUTPUT")
+	baseFile := os.Getenv("BASE")
+	nextFile := os.Getenv("NEXT")
+	var args []string
+	if baseFile == "" {
+		log.Printf("Missing required parameter BASE parameter to parse the results. Please set this to a filepath of the benchmark results")
+		return fmt.Errorf("missing required parameter BASE parameter to parse the results. Please set this to a filepath of the benchmark results")
+	} else {
+		args = append(args, baseFile)
+	}
+	if nextFile == "" {
+		log.Printf("Missing NEXT parameter, we are not going to compare results")
+	} else {
+		args = append(args, nextFile)
+	}
+
+	_, err := runCommand(ctx, nil, "benchstat", outputFile, args...)
+
+	var goTestErr *exec.ExitError
+	switch {
+	case goTestErr == nil:
+		return nil
+	case errors.As(err, &goTestErr):
+		return fmt.Errorf("failed to execute benchstat command: %w", err)
+	default:
+		return fmt.Errorf("failed to execute benchstat command!! %w", err)
+	}
+
+}
+
+// runCommand is executing a command that is represented by cmd.
+// when defining an outputFile it will write the stdErr, stdOut of that command to the output file
+// otherwise it will capture it to stdErr, stdOut of the console used.
+func runCommand(ctx context.Context, env map[string]string, cmd string, outputFile string, args ...string) (*exec.Cmd, error) {
```

Comment:
Is there a reason we're not using the mage [sh](https://pkg.go.dev/github.com/magefile/mage@v1.14.0/sh) libraries here? Might remove some of the yak shaving.

Reply:
Great suggestion, I really don't know why. I recall I tried to use it but I had some issues that I cannot reproduce anymore, so the problem was unrelated to the use of `sh` that pushed me to use the `exec` directly.